bridges/snowbridge/primitives/core/src/lib.rs

```diff
@@ -176,38 +161,63 @@ pub struct AssetMetadata {
 	pub decimals: u8,
 }
 
+pub type AgentId = H256;
+
+/// Creates an AgentId from a Location. An AgentId is a unique mapping to a Agent contract on
+/// Ethereum which acts as the sovereign account for the Location.
+pub type AgentIdOf =
+	HashedDescription<AgentId, (DescribeHere, DescribeFamily<DescribeAllTerminal>)>;
+
 pub type TokenId = H256;
 
 /// Convert a token location to a stable ID that can be used on the Ethereum side
-pub type TokenIdOf = HashedDescription<TokenId, DescribeGlobalPrefix<DescribeInner>>;
+pub type TokenIdOf =
+	HashedDescription<TokenId, DescribeGlobalPrefix<DescribeFamily<DescribeToken>>>;
 
+pub struct DescribeHere;
+impl DescribeLocation for DescribeHere {
+	fn describe_location(l: &Location) -> Option<Vec<u8>> {
+		match l.unpack() {
+			(0, []) => Some(Vec::<u8>::new().encode()),
+			_ => None,
+		}
+	}
+}
 pub struct DescribeGlobalPrefix<DescribeInterior>(sp_std::marker::PhantomData<DescribeInterior>);
 impl<Suffix: DescribeLocation> DescribeLocation for DescribeGlobalPrefix<Suffix> {
 	fn describe_location(l: &Location) -> Option<Vec<u8>> {
 		match (l.parent_count(), l.first_interior()) {
-			(1, Some(GlobalConsensus(network))) => {
+			(_, Some(GlobalConsensus(network))) => {
 				let tail = l.clone().split_first_interior().0;
 				let interior = Suffix::describe_location(&tail.into())?;
-				Some((b"pna", network, interior).encode())
+				Some((b"PNA", network, interior).encode())
 			},
 			_ => None,
 		}
 	}
 }
 
-pub struct DescribeInner;
-impl DescribeLocation for DescribeInner {
+pub struct DescribeToken;
+impl DescribeLocation for DescribeToken {
 	fn describe_location(l: &Location) -> Option<Vec<u8>> {
 		match l.unpack().1 {
 			[] => Some(Vec::<u8>::new().encode()),
-			[Parachain(id)] => Some((*id).encode()),
-			[Parachain(id), PalletInstance(instance)] => Some((*id, *instance).encode()),
-			[Parachain(id), PalletInstance(instance), GeneralIndex(index)] =>
-				Some((*id, *instance, *index).encode()),
-			[Parachain(id), PalletInstance(instance), GeneralKey { data, .. }] =>
-				Some((*id, *instance, *data).encode()),
-			[Parachain(id), GeneralIndex(index)] => Some((*id, *index).encode()),
-			[Parachain(id), GeneralKey { data, .. }] => Some((*id, *data).encode()),
+			[GeneralIndex(index)] => Some((*index).encode()),
+			[GeneralKey { data, .. }] => Some((*data).encode()),
+			[AccountKey20 { key, .. }] => Some((*key).encode()),
+			[AccountId32 { id, .. }] => Some((*id).encode()),
+
+			// Pallet
+			[PalletInstance(instance)] => Some((*instance).encode()),
+			[PalletInstance(instance), GeneralIndex(index)] => Some((*instance, *index).encode()),
+			[PalletInstance(instance), GeneralKey { data, .. }] =>
+				Some((*instance, *data).encode()),
+
+			[PalletInstance(instance), AccountKey20 { key, .. }] =>
+				Some((*instance, *key).encode()),
+			[PalletInstance(instance), AccountId32 { id, .. }] => Some((*instance, *id).encode()),
```

Comment:
These cases are all exploitable by collision attacks. The name of the junction should be included in the preimage, as is done for example by `DescribeAllTerminal`.